scripts/midi_remap.lua

```diff
@@ -0,0 +1,103 @@
+ardour {
+    ["type"]    = "dsp",
+    name        = "MIDI Remap Notes",
+    category    = "Utility",
+    license     = "MIT",
+    author      = "Alby Musaelian",
+    description = [[Map arbitrary MIDI notes to others. Affects Note On/Off and polyphonic key pressure.]]
+}
+
+-- The number of remapping pairs to allow. Increasing this (at least in theory)
+-- decreases performace, so it's set fairly low as a default. The user can
+-- increase this if they have a need to.
+N_REMAPINGS = 10
+
+function dsp_ioconfig ()
+    return { { midi_in = 1, midi_out = 1, audio_in = 0, audio_out = 0}, }
+end
+
+function dsp_params ()
+    local note_names = {"C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"}
+
+    local map_scalepoints = {}
+    for note=0,127 do
+        local octave = -5 + math.floor(note / 12)
+        local name = note_names[(note % 12) + 1]
+        map_scalepoints[string.format("%03d (%s%d)", note, name, octave)] = note
+    end
+
+    local map_params = {}
+
+    i = 1
+    for mapnum = 1,N_REMAPINGS do
+        -- From and to
+        for _,name in pairs({"| #" .. mapnum .. "  Map note", "|__   to"}) do
+            map_params[i] = {
+                ["type"] = "input",
+                name = name,
+                min = 0,
+                max = 127,
+                default = 0,
+                integer = true,
+                enum = true,
+                scalepoints = map_scalepoints
+            }
+            i = i + 1
+        end
+    end
+
+    return map_params
+end
+
+function dsp_init(samplerate)
+    -- Init translation table, reserve memory
+    translation_table = {}
+    for i = 0,127 do
+        translation_table[i] = i
+    end
+end
+
+
+function dsp_run (_, _, n_samples)
+    assert (type(midiin) == "table")
+    assert (type(midiout) == "table")
+    local cnt = 1;
+
+    function tx_midi (time, data)
+        midiout[cnt] = {}
+        midiout[cnt]["time"] = time;
+        midiout[cnt]["data"] = data;
+        cnt = cnt + 1;
+    end
+
+    -- We build the translation table every buffer because, as far as I can tell,
+    -- there's no way to only rebuild it when the parameters have changed.
+    -- As a result, it has to be updated every buffer for the parameters to have
+    -- any effect.
```

Comment:
You could store it globally and only update it when a parameter has changed. Not sure if it's worth it though.

An alternative is to not initialize the table with a unity matrix but use Lua's sparse array. 

```
  local translation_table = {}
  local ctrl = CtrlPorts:array()
  for i=1,N_REMAPINGS*2,2 do
    translation_table[ctrl[i]] = ctrl[i + 1]
  end
-- .....
  d[2] = translation_table[d[2]] or d[2]
```

Reply:
Is there a callback for when the parameters have changed? If not, I would have to store the past parameter values and compare them, which takes nearly as many operations as rebuilding the table and introduces an extra branch, so I doubt it would actually change performance (I know that's C-level nitpicking, but either way I doubt the differences are measurable at this level).

I opted against the sparse array since my research indicated that tables indexed with contiguous positive integers were implemented under the hood as C-style arrays, and as a result would hopefully be efficient... but since it looks like most Lua versions [do a hash table anyway](https://stackoverflow.com/questions/26843589/how-does-lua-5-x-represent-sparse-arrays), maybe it's better to just clean up the code and use one.

Reply:
Whoops. Actually, the sparse array won't help: say there is a mapping from 2 -> 3 that is later changed to 5 -> 3, for example. If `translation_table` isn't reinitialized to the identity, now *both* 2 and 5 will map to 3... I think I'll leave it as is.

(I'm also pretty sure it's already a global -- omitting `local` in the declaration is sufficient, right?)

Reply:
> Is there a callback for when the parameters have changed? If not, I would have to store the past parameter values and compare them

yes, you would have to store and compare. It's not worth it.

>  If translation_table isn't reinitialized to the identity, ..

`local translation_table = {}` initializes it -- you can also get rid of the the `dsp_init` function and global definition.